### ISSUE TITLE
"prefer-const" rule to discourage unnecessary use of "let"

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#491]
I'd like a new rule that prevents declaring variables as `let` _unless that reference is later mutated via a left-hand-side assignment_. `const` should be used instead for better readability if you intend to never re-assign the variable.

Some code examples:

``` ts
// bad
for (let i of [1, 2, 3]) {
  //
}

// good
for (const i of [1, 2, 3]) {
  //
}

// bad
let {createReadStream} = require("fs");

// good, unless the recently-introduced "no-require-imports" rule is also enabled
const {createReadStream} = require("fs");
```

**Comment by thasner** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
@weswigham has done this for us. I think he'll be able to send out a PR eventually.

**Comment by weswigham** [comment]
[Yeah, I've worked on it.](https://github.com/Microsoft/TypeScript/pull/5531)

**Comment by buzinas** [comment]
@weswigham any news on this?

**Comment by Niondir** [comment]
+1 - eslint has this rule

**Comment by blakehawkins** [comment]
It appears that the mentioned PR was merged into Microsoft/typescript. Has anyone had this rule working successfully?

**Comment by scottcheng** [comment]
It's in [tslint-microsoft-contrib](https://github.com/Microsoft/tslint-microsoft-contrib/blob/f8b057615332046dcd52cf6e6e5ca4357fd592f3/src/preferConstRule.ts) and it's working for me (for the most part).

**Comment by westy92** [comment]
I just got this rule working.

```npm install --save-dev tslint-microsoft-contrib```

`tslint.json`:
```json
{
  "extends": ["tslint:recommended"],
  "rules": {
     "prefer-const": true
  },
  "rulesDirectory": [
     "node_modules/tslint-microsoft-contrib"
  ]
}
```

**Comment by adidahiya** [comment]
assigning to @nchen63 to tackle after the 4.0 release. here's a WIP branch I started a while ago... https://github.com/palantir/tslint/tree/wip-prefer-const
